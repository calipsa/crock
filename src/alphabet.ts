export default '0123456789abcdefghjkmnpqrstvwxyz'
