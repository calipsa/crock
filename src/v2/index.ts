export {
  encodeCamera,
  encodeSite,
  encodeSubject,
} from './encode'

export {
  decode6crock,
  decode10crock,
} from './decode'
