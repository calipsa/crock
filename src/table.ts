import alphabet from './alphabet'

// tslint:disable-next-line:interface-over-type-literal
type Table = {
  [letter: string]: number,
}

const alias = {
  o: 0,
  i: 1,
  l: 1,
}

const table: Table = {}
// Invert 'alphabet'
let i = 0
for (const letter of alphabet) {
  table[letter] = i++
}
// Splice in 'alias'
for (const [key, val] of Object.entries(alias)) {
  table[key] = table[val]
}

export default Object.freeze(table)
