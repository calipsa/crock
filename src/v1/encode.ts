import alphabet from '../alphabet'

function hexToOctet(hex: string) {
  const byteList: number[] = []
  const len = hex.length
  for (let i = 0; i < len; i += 2) {
    const hexPair = hex.substring(i, i + 2)
    const octet = Number.parseInt(hexPair, 16)
    byteList.push(octet)
  }
  return byteList
}

function encode(version: number, [a, b, c, d, e, f]: number[]) {
  const values = [
    (version << 3) | (a >>> 5),
    a & 0x1f,
    b >>> 3,
    (b & 7) << 2 | (c >>> 6),
    (c >>> 1) & 0x1f,
    (c & 1) << 4 | (d >>> 4),
    (d & 0xf) << 1 | (e >>> 7),
    (e >>> 2) & 0x1f,
    (e & 3) << 3 | (f >>> 5),
    f & 0x1f,
  ]
  return values
    .map(value => alphabet[value])
    .join('')
}

export function encodeUuid(uuid: string) {
  const first48bits = `${uuid.substring(0, 8)}${uuid.substring(9, 13)}`
  const octetList = hexToOctet(first48bits)
  return encode(0, octetList)
}
