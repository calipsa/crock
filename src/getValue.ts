/*
 * https://en.wikipedia.org/wiki/Email_box#Mailbox_names
 * https://en.wikipedia.org/wiki/Universally_unique_identifier#Format
 * https://en.wikipedia.org/wiki/Base32
 * http://www.crockford.com/wrmg/base32.html
 * https://github.com/latentflip/base32-crockford-browser/blob/master/lib/base32.js
 */

import table from './table'

export default (letter: string) => {
  // eslint-disable-next-line no-prototype-builtins
  if (!table.hasOwnProperty(letter)) {
    throw new Error(`invalid encoding: "${letter}"`)
  }
  return table[letter]
}
