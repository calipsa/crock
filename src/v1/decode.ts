import table from '../table'

import octetToHex from '../utils/octetToHex'

function decode(letters: string) {
  const [a, b, c, d, e, f, g, h, i, j] = [...letters].map(l => table[l])
  // const version = a >>> 3
  return [
    ((a & 7) << 5) | b,
    c << 3 | (d >>> 2),
    (d & 3) << 6 | (e << 1) | (f >>> 4),
    (f & 0xf) << 4 | g >>> 1,
    (g & 1) << 7 | h << 2 | i >>> 3,
    (i & 7) << 5 | j,
  ]
}

export default (crock: string) => {
  const uuid = octetToHex(decode(crock))
  return `${uuid.slice(0, 8)}-${uuid.slice(8, 12)}`
}
