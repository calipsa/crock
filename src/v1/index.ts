import alphabet from '../alphabet'
import table from '../table'

function hexToOctet(hex: string) {
  const byteList: number[] = []
  const len = hex.length
  for (let i = 0; i < len; i += 2) {
    const hexPair = hex.substring(i, i + 2)
    const octet = Number.parseInt(hexPair, 16)
    byteList.push(octet)
  }
  return byteList
}

/*
 * quicker or slower?
 */
/*
function octetToHex (list) {
  const hex = []
  for (let o of list) {
    let hexPair = o.toString(16)
    if (hexPair.length < 2)
      hexPair = '0' + hexPair
    hex.push(hexPair)
  }
  return hex.join('')
}
*/

function octetToHex(list: number[]) {
  const hex: string[] = []
  for (const o of list) {
    const pair = [o >>> 4, o & 0xf].map(v => v.toString(16))
    hex.push(...pair)
  }
  return hex.join('')
}

function encode(version: number, [a, b, c, d, e, f]: number[]) {
  const values = [
    (version << 3) | (a >>> 5),
    a & 0x1f,
    b >>> 3,
    (b & 7) << 2 | (c >>> 6),
    (c >>> 1) & 0x1f,
    (c & 1) << 4 | (d >>> 4),
    (d & 0xf) << 1 | (e >>> 7),
    (e >>> 2) & 0x1f,
    (e & 3) << 3 | (f >>> 5),
    f & 0x1f,
  ]
  return values
    .map(value => alphabet[value])
    .join('')
}

function decode(letters: string) {
  const values: number[] = []
  for (const l of letters) {
    values.push(table[l])
  }
  const [a, b, c, d, e, f, g, h, i, j] = values
  // const version = a >>> 3
  const bytes = [
    ((a & 7) << 5) | b,
    c << 3 | (d >>> 2),
    (d & 3) << 6 | (e << 1) | (f >>> 4),
    (f & 0xf) << 4 | g >>> 1,
    (g & 1) << 7 | h << 2 | i >>> 3,
    (i & 7) << 5 | j,
  ]
  return bytes
}

export function encodeUuid(uuid: string) {
  const first48bits = `${uuid.substring(0, 8)}${uuid.substring(9, 13)}`
  const octetList = hexToOctet(first48bits)
  return encode(0, octetList)
}

export function decodeUuid(crock: string) {
  const uuid = octetToHex(decode(crock))
  return `${uuid.substring(0, 8)}-${uuid.substring(8, 12)}`
}
