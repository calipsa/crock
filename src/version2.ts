import alphabet from './alphabet'
import getValue from './getValue'

function decodeCrock(letters: string) {
  let bin = 0
  for (const l of letters) {
    bin <<= 5
    bin |= getValue(l)
  }
  return bin
}

const decode4crockToHex = (letters: string) =>
  decodeCrock(letters).toString(16).padStart(5, '0')

const decode8crockToHex = (letters: string) =>
  `${decode4crockToHex(letters.slice(0, 4))}${decode4crockToHex(letters.slice(4, 8))}`

function encode7hex(version: number, hexs: string) {
  let bin = Number.parseInt(hexs, 16)
  let crockStr = ''
  for (let i = 0; i < 5; i++, bin >>>= 5) {
    crockStr = `${alphabet[bin & 31]}${crockStr}`
  }

  return `${alphabet[version << 3 | bin]}${crockStr}`
}

function encode5hex(hex: string) {
  let bin = Number.parseInt(hex, 16)
  let crockStr = ''
  for (let i = 0; i < 4; i++, bin >>>= 5) {
    crockStr = `${alphabet[bin & 0x1f]}${crockStr}`
  }

  return crockStr
}

const encode10hex = (hex: string) =>
  `${encode5hex(hex.slice(0, 5))}${encode5hex(hex.slice(5, 10))}`

export function encodeSubject(uuid: string) {
  const first7hex = uuid.slice(0, 7)
  return encode7hex(0, first7hex)
}

function encodeMailbox(uuid: string, isNvr: boolean) {
  const version = 0
  const fst12hex = `${uuid.slice(0, 8)}${uuid.slice(9, 13)}`
  const [fstHex, sndHex, ...last10hex] = fst12hex
  const last8crock = encode10hex(last10hex.join(''))
  const fst2bits = version << 1 | (+isNvr)
  const fst10bits = fst2bits << 8 | Number.parseInt(fstHex + sndHex, 16)
  const fst5bits = fst10bits >>> 5
  const snd5bits = fst10bits & 0x1f
  return `${alphabet[fst5bits]}${alphabet[snd5bits]}${last8crock}`
}

export const encodeCamera = (uuid: string) =>
  encodeMailbox(uuid, false /* not an NVR */)

export const encodeSite = (uuid: string) =>
  encodeMailbox(uuid, true /* is an NVR */)

/*
 * first two bits were planned for versioning, but with the new
 * NVR requirement the second bit is used as an isNvr flag leaving
 * just the first bit for versioning/reserved.
 */
export function decode10crock(crock: string) {
  if (crock.length !== 10) {
    throw new Error('unknown encoding - length different from 10 characters')
  }

  const letters = crock.toLowerCase()
  const fst5bits = getValue(letters[0])
  const reserved = fst5bits >>> 4
  if (reserved) {
    throw new Error('unknown encoding')
  }

  const snd5bits = getValue(letters[1])
  const fst8bits = (fst5bits & 7) << 5 | snd5bits
  const fst2Hex = fst8bits.toString(16).padStart(2, '0')
  const last10Hex = decode8crockToHex(letters.slice(2))

  return {
    isNvr: !!(fst5bits >>> 3 & 1),
    uuid: `${fst2Hex}${last10Hex.slice(0, 6)}-${last10Hex.slice(6)}`,
  }
}

export function decode6crock(crock: string) {
  if (crock.length !== 6) {
    throw new Error('unknown encoding - length different from 10 characters')
  }

  const letters = crock.toLowerCase()
  let bin = decodeCrock(letters)
  const version = bin >>> 28 // two first bits for version
  if (version) {
    throw new Error('unknown encoding version')
  }

  bin &= 0xfffffff
  return {
    uuid: bin.toString(16).padStart(7, '0'),
  }
}
