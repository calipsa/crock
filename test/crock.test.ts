import * as Uuid from 'uuid/v4'

import * as OldCrock from '../src/v1'
import * as Crock from '../src/index'

describe('Testing crock', () => {
  it('Should throw on bad UUID argument', () => {
    const badUuid = 'foobarbaz'
    expect(() => Crock.encodeCamera(badUuid)).toThrowErrorMatchingSnapshot()
    expect(() => Crock.encodeSite(badUuid)).toThrowErrorMatchingSnapshot()
    expect(() => Crock.encodeSubject(badUuid)).toThrowErrorMatchingSnapshot()
  })

  it('Should throw on bad string length argument', () => {
    expect(() => Crock.decode10crock('not10characters&badenc')).toThrowErrorMatchingSnapshot()
    expect(() => Crock.decode6crock('not6characters&badenc')).toThrowErrorMatchingSnapshot()

    expect(() => Crock.decode10crock('not10characters')).toThrowErrorMatchingSnapshot()
    expect(() => Crock.decode6crock('not6characters')).toThrowErrorMatchingSnapshot()

    expect(() => Crock.decode10crock('bad_encod+')).toThrowErrorMatchingSnapshot()
    expect(() => Crock.decode6crock('bad_e$')).toThrowErrorMatchingSnapshot()
  })

  it('Should return false for every camera mailbox UUID codec', () => {
    for (let i = 0; i < 200; ++i) {
      const uuidSrc = Uuid()
      const camera = Crock.encodeCamera(uuidSrc)
      const { isNvr, uuid } = Crock.decode10crock(camera)
      expect(isNvr).toBe(false)
      expect(uuid).toBe(uuidSrc.slice(0, 13))
    }
  })

  it('Should return true for every site mailbox UUID codec', () => {
    for (let i = 0; i < 200; ++i) {
      const uuidSrc = Uuid()
      const site = Crock.encodeSite(uuidSrc)
      const { isNvr, uuid } = Crock.decode10crock(site)
      expect(isNvr).toBe(true)
      expect(uuid).toBe(uuidSrc.slice(0, 13))
    }
  })

  it('Should check UUID of every email subject UUID codec', () => {
    for (let i = 0; i < 200; ++i) {
      const uuidSrc = Uuid()
      const subject = Crock.encodeSubject(uuidSrc)
      const { uuid } = Crock.decode6crock(subject)
      expect(uuid).toBe(uuidSrc.slice(0, 7))
    }
  })

  it('Should check if new mailbox codec behaves as old', () => {
    for (let i = 0; i < 200; ++i) {
      const uuidSrc = Uuid()
      const cameraNew = Crock.encodeCamera(uuidSrc)
      const cameraOld = OldCrock.encodeUuid(uuidSrc)
      expect(cameraNew).toBe(cameraOld)
      const { isNvr, uuid } = Crock.decode10crock(cameraNew)
      const uuidOld = OldCrock.decodeUuid(cameraNew)
      expect(isNvr).toBe(false)
      expect(uuidOld).toBe(uuid)
      expect(uuid).toBe(uuidSrc.slice(0, 13))
    }
  })
})
