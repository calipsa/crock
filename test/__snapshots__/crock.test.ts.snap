// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing crock Should throw on bad UUID argument 1`] = `"Expected: a valid UUID. Got: foobarbaz"`;

exports[`Testing crock Should throw on bad UUID argument 2`] = `"Expected: a valid UUID. Got: foobarbaz"`;

exports[`Testing crock Should throw on bad UUID argument 3`] = `"Expected: a valid UUID. Got: foobarbaz"`;

exports[`Testing crock Should throw on bad string length argument 1`] = `"Expected: string 10 characters long. Got: not10characters&badenc"`;

exports[`Testing crock Should throw on bad string length argument 2`] = `"Expected: string 6 characters long. Got: not6characters&badenc"`;

exports[`Testing crock Should throw on bad string length argument 3`] = `"Expected: string 10 characters long. Got: not10characters"`;

exports[`Testing crock Should throw on bad string length argument 4`] = `"Expected: string 6 characters long. Got: not6characters"`;

exports[`Testing crock Should throw on bad string length argument 5`] = `"Expected: correctly encoded string. Got: bad_encodi"`;

exports[`Testing crock Should throw on bad string length argument 6`] = `"Expected: correctly encoded string. Got: bad_en"`;
